### PR TITLE
P0: Add input validation and fix authorization gaps

### PR DESCRIPTION
**Pull request by jame581**
## Summary

- Fix authorization gaps on single-entity GET endpoints (GetProject, GetTrackedTask)
- Add Data Annotation validation attributes to all create/update DTOs
- Add custom MinDuration attribute for TimeSpan > 0 validation
- Add EndDate >= StartDate validation in UpdateTrackedTask
- 3 new tests covering authorization and validation edge cases

## Security Fixes

- **GetProject(id)** now returns 404 if the project doesn't belong to the requesting user (was returning any user's project)
- **GetTrackedTask(id)** now returns 404 if the task doesn't belong to the requesting user (was returning any user's task)
- Both return NotFound (not Unauthorized) to avoid revealing resource existence

## Validation Added

| DTO | Field | Rules |
|-----|-------|-------|
| CreateProjectDto | Name | Required, 2-100 chars |
| UpdateProjectDto | ProjectId | Required |
| UpdateProjectDto | Name | Required, 2-100 chars |
| CreateTrackedTaskDto | Duration | Required, > 0 (custom MinDuration) |
| UpdateTrackedTaskDto | TaskId | Required |
| UpdateTrackedTaskDto | Name | Required, 2-50 chars |
| UpdateTrackedTaskDto | StartDate | Required |
| UpdateTrackedTask | EndDate | Must be >= StartDate (controller check) |

## Test plan

- [x] `dotnet build Timinute.sln` — 0 errors
- [x] `dotnet test` — 50/50 passing (47 existing + 3 new)
- [x] `Get_Project_Another_User_Returns_NotFound_Test` — verifies ownership check
- [x] `Get_TrackedTask_Another_User_Returns_NotFound_Test` — verifies ownership check
- [x] `Update_TrackedTask_EndDate_Before_StartDate_Returns_BadRequest_Test` — verifies date validation

### CODE REVIEW

File: Timinute/Server/Controllers/TrackedTaskController.cs [hunk 1]
```diff
@@ -160,6 +160,12 @@ public async Task<ActionResult<TrackedTaskDto>> UpdateTrackedTask([FromBody] Upd
             if (updatedTrackedTask.EndDate.HasValue)
             {
                 updatedTrackedTask.EndDate = updatedTrackedTask.EndDate.Value.ToUniversalTime();
+
+                if (updatedTrackedTask.EndDate.Value <= updatedTrackedTask.StartDate)
+                {
+                    return BadRequest("End date must be after start date.");
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/TrackedTaskController.cs]: Fixed in 911c617 — updated error message to 'End date must be strictly after start date.' to clarify the <= comparison rejects zero-duration. This is intentional to stay consistent with MinDuration > 0 on create.

File: Timinute/Shared/Dtos/TrackedTask/CreateTrackedTaskDto.cs [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
 {
     public class CreateTrackedTaskDto
     {
         [Required]
-        [StringLength(50, ErrorMessage = "Name of project is too long.", MinimumLength = 2)]
+        [StringLength(50, ErrorMessage = "Name of task is too long.", MinimumLength = 2)]
         public string Name { get; set; } = null!;
 
         [Required]
         public DateTime StartDate { get; set; }
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Shared/Dtos/TrackedTask/CreateTrackedTaskDto.cs]: Fixed in 911c617 — changed StartDate to DateTime? so [Required] properly rejects missing values instead of silently accepting DateTime.MinValue. Updated controller to use .Value after validation guarantees non-null.

File: Timinute/Shared/Dtos/TrackedTask/UpdateTrackedTaskDto.cs [hunk 3]
```diff
@@ -1,14 +1,24 @@
...
     {
+        [Required]
         public string TaskId { get; set; } = null!;
+
+        [Required]
+        [StringLength(50, ErrorMessage = "Name of task is too long.", MinimumLength = 2)]
         public string Name { get; set; } = null!;
+
+        [Required]
         public DateTime StartDate { get; set; }
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Shared/Dtos/TrackedTask/UpdateTrackedTaskDto.cs]: Fixed in 911c617 — same fix applied, StartDate is now DateTime? with [Required].

File: Timinute/Shared/Dtos/TrackedTask/UpdateTrackedTaskDto.cs [hunk 4]
```diff
@@ -1,14 +1,24 @@
...
 
 namespace Timinute.Shared.Dtos.TrackedTask
 {
     public class UpdateTrackedTaskDto
     {
+        [Required]
         public string TaskId { get; set; } = null!;
+
+        [Required]
+        [StringLength(50, ErrorMessage = "Name of task is too long.", MinimumLength = 2)]
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Shared/Dtos/TrackedTask/UpdateTrackedTaskDto.cs]: Fixed in 911c617 — updated to 'Name of task must be between 2 and 50 characters long.' on both Create and Update DTOs.

File: Timinute/Shared/Validators/MinDurationAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        }
+
+        protected override ValidationResult? IsValid(object? value, ValidationContext validationContext)
+        {
+            if (value is TimeSpan duration && duration <= TimeSpan.Zero)
+            {
+                return new ValidationResult(ErrorMessage);
+            }
+
+            return ValidationResult.Success;
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Shared/Validators/MinDurationAttribute.cs:23]: Fixed in 911c617 — MinDurationAttribute now returns a validation error for null and non-TimeSpan values instead of silently passing.

File: Timinute/Shared/Dtos/TrackedTask/CreateTrackedTaskDto.cs [hunk 6]
```diff
@@ -1,17 +1,20 @@
...
 using Timinute.Shared.Dtos.Project;
+using Timinute.Shared.Validators;
 
 namespace Timinute.Shared.Dtos.TrackedTask
 {
     public class CreateTrackedTaskDto
     {
         [Required]
-        [StringLength(50, ErrorMessage = "Name of project is too long.", MinimumLength = 2)]
+        [StringLength(50, ErrorMessage = "Name of task is too long.", MinimumLength = 2)]
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Shared/Dtos/TrackedTask/CreateTrackedTaskDto.cs]: Fixed in 911c617 — same message fix applied.

File: Timinute/Server/Controllers/TrackedTaskController.cs [hunk 7]
```diff
@@ -93,8 +93,8 @@ public async Task<ActionResult<TrackedTaskDto>> CreateTrackedTask([FromBody] Cre
 
             var newTrackedTask = mapper.Map<TrackedTask>(trackedTask);
             newTrackedTask.UserId = userId;
-            newTrackedTask.StartDate = trackedTask.StartDate.ToUniversalTime();
-            newTrackedTask.EndDate = trackedTask.StartDate + newTrackedTask.Duration;
+            newTrackedTask.StartDate = trackedTask.StartDate!.Value.ToUniversalTime();
+            newTrackedTask.EndDate = trackedTask.StartDate.Value + newTrackedTask.Duration;
```
[the remark this answers is not in the document]
Comment (jame581) [reply Timinute/Server/Controllers/TrackedTaskController.cs]: Good catch — real bug. Fixed in 12ae20b. EndDate is now computed from the UTC-converted StartDate variable instead of the raw DTO value, keeping both timestamps in UTC consistently.

File: docs/superpowers/plans/2026-04-01-p0-validation-authorization.md [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+updatedTrackedTask.StartDate = updatedTrackedTask.StartDate.ToUniversalTime();
+
+if (updatedTrackedTask.EndDate.HasValue)
+{
+    updatedTrackedTask.EndDate = updatedTrackedTask.EndDate.Value.ToUniversalTime();
+
+    if (updatedTrackedTask.EndDate.Value < updatedTrackedTask.StartDate)
+    {
+        return BadRequest("End date must be after start date.");
+    }
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/plans/2026-04-01-p0-validation-authorization.md:424]: Fixed in 12ae20b — updated plan to use <= check and 'strictly after' message matching the implementation.

File: docs/superpowers/plans/2026-04-01-p0-validation-authorization.md [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+- [ ] **Step 5: Run all tests**
+
+Run: `dotnet test`
+Expected: 50/50 pass
+
+- [ ] **Step 6: Commit**
+
+```bash
+git add Timinute/Server/Controllers/TrackedTaskController.cs Timinute/Server.Tests/Controllers/TrackedTaskControllerTest.cs
+git commit -m "feat: validate EndDate >= StartDate in UpdateTrackedTask"
```
[the remark this answers is not in the document]
Comment (jame581) [reply docs/superpowers/plans/2026-04-01-p0-validation-authorization.md]: Fixed in 12ae20b — commit message in plan now says 'EndDate > StartDate' matching the strict comparison.
